src/applications/applications.service.ts

```diff
@@ -22,505 +24,689 @@ type BenefitDefinition = {
-      return false;
-    }
-  }
+			// Only allow safe characters
+			if (!/^[\d\s+\-*/().]+$/.test(safeExpr))
+				throw new Error('Unsafe formula');
+
+			return Function(`"use strict"; return (${safeExpr})`)(); // evaluated safely
+		} catch (err) {
+			console.error('Formula evaluation error:', err.message);
+			return 0;
+		}
+	}
+
+	/**
+	 * Limited expression evaluator supporting basic comparison logic.
+	 */
+	evaluateIfExpr(expr: string, context: ApplicationData): boolean {
+		try {
+			// Basic parser for comparison operators
+			const comparisons = expr.match(
+				/([a-zA-Z_][a-zA-Z0-9_]*)\s*([=!<>]+)\s*(true|false|\d+|"[^"]*"|'.*?')/g,
+			);
+
+			if (!comparisons) return false;
+
+			return comparisons.every((part) => {
+				const [, key, op, rawVal] =
+					part.match(/([a-zA-Z_][a-zA-Z0-9_]*)\s*([=!<>]+)\s*(.*)/) || [];
+				let actual = context[key];
+				let expected: any = rawVal;
+
+				if (expected === 'true') expected = true;
+				else if (expected === 'false') expected = false;
+				else if (!isNaN(Number(expected))) expected = Number(expected);
+				else expected = expected.replace(/^['"]|['"]$/g, '');
+
+				switch (op) {
+					case '===':
+						return actual === expected;
+					case '!==':
+						return actual !== expected;
+					case '>':
+						return actual > expected;
+					case '<':
+						return actual < expected;
+					case '>=':
+						return actual >= expected;
+					case '<=':
+						return actual <= expected;
+					default:
+						return false;
+				}
+			});
+		} catch (err) {
+			console.error('Expression evaluation error:', err.message);
+			return false;
+		}
+	}
+
+	async checkEligibility(applicationId: number, req) {
+		const application = await this.findOne(applicationId); // Fetch the application by ID
+
+		if (!application) {
+			throw new NotFoundException(
+				`Application with ID ${applicationId} not found`,
+			);
+		}
+
+		const benefitDefinition = await this.benefitsService.getBenefitsById(
+			`${application.benefitId}`,
+		);
+		if (!benefitDefinition?.data) {
+			throw new NotFoundException(
+				`Benefit with ID ${application.benefitId} not found`,
+			);
+		}
+		const strictCheck = req?.query?.strictCheck === 'true';
+		const eligibilityResult = await this.checkAndFormatEligibility(
+			benefitDefinition,
+			application,
+			strictCheck,
+		);
+		let eligibilityStatus = 'ineligible'; // Default status
+		if (eligibilityResult?.eligibleUsers?.length > 0) {
+			eligibilityStatus = 'eligible'; // Set to eligible if any users are eligible default we sending one application here
+		}
+		await this.update(applicationId, {
+			eligibilityStatus,
+			eligibilityResult: eligibilityResult,
+			eligibilityCheckedAt: new Date(),
+		});
+
+		return eligibilityResult;
+	}
+
+	/**
+	 * Prepares eligibility rules and application profile, then calls the eligibility API.
+	 * Returns the eligibility check result.
+	 */
+	async checkAndFormatEligibility(
```

Comment:
@milindSansare The function should have only one responsibility princial i.e one function should do only one task.

We can split the format and check part as different function. You have created `checkApplicationEligibility` which is fine and good but this should not be called from `checkAndFormatEligibility` rather it should be called from the parent function.

after spliting the functions rename `checkAndFormatEligibility` => `getEligibilityRules`